### ISSUE TITLE
Promise being resolved with wrong type

### ISSUE THREAD

**Issue by 0xfede** [microsoft/TypeScript#9991]
**TypeScript Version:**  1.8.10 and 2.0.0 beta

**Code**

``` ts
function getPromise(): Promise<string> {
  return new Promise(resolve => {
    resolve(1234); // Should fail as 'abcd' is not a number
  });
}

getPromise().then(s => {
  console.log(s.length);
});
```

**Expected behaviour:** A compile error on line 3

**Actual behaviour:**  The program prints `undefined`

If I copy the type definition of `resolve` from the interface `PromiseConstructor` (in lib.es6.d.ts), I get the expected behaviour:

``` ts
function getPromise(): Promise<string> {
  return new Promise((resolve: (value?: string | PromiseLike<string>) => void) => {
    resolve(1234); // As expected I get an error
    // TS2345: Argument of type 'number' is not assignable to parameter of type 'string | PromiseLike<string>' Type 'number' is not assignable to type 'PromiseLike<string>'.
  });
}

getPromise().then(s => {
  console.log(s.length);
});
```

**Comment by jeffreymorlan** [comment]
It's easier to specify the type argument: `new Promise<string>(resolve => {...})`

Unfortunately, `new Promise(resolve => {...})` without any type annotations always has a type of `Promise<{}>` due to https://github.com/Microsoft/TypeScript/issues/5254, which is assignable to any other Promise due to https://github.com/Microsoft/TypeScript/issues/6407.

**Comment by badescuga** [comment]
I'm also seeing this :) Not sure why the time is not being respected.

**Comment by waratah** [comment]
@badescuga   any is a catch all and will totally break type safety.   Got me last week.   The more definitions you introduce in your code the better.   You can also use noImplicitAny to pick up on these errors.

**Comment by RyanCavanaugh** [comment]
This has been fixed

**Comment by acashjos** [comment]
**TypeScript Version:**  Version 3.6.3

Just a reminder, while the fix works for most practical cases, it still doesn't detect type inconsistencies for  `Promise<Object >` , `Promise<{}>`  or any interface defined as `{}`
 ```js
interface Message { }
function getPromise(): Promise<Message> {
    return new Promise(resolve => {
        resolve(1234); // Should fail as 'abcd' is not a Message
    });
}

getPromise().then(s => {
    console.log(s.length);  // fails as length is not a property of Message
});
```
